File: sdk-go/dex/worker.go
// Legacy Materials in this file remain under their original licenses.
// See LEGACY_NOTICES.md.

// Modifications Copyright (c) 2026 Super Durable, Inc.
//
// Modifications after the Legacy Cutoff are licensed under the
// Super Durable Source License 1.0.
// Legacy Materials remain under their original licenses.
// See LICENSE and LEGACY_NOTICES.md.

package dex

import (
	"context"
	"errors"
	"fmt"
	"net"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/superdurable/dex/blob-cache-go/blobcache"
	"github.com/superdurable/dex/sdk-go/gen/dexpb"
	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials/insecure"
)

const (
	defaultWorkerBindAddress         = ":8803"
	defaultFlowServiceTarget         = "localhost:8801"
	defaultAttributeIndexSyncTimeout = 2 * time.Minute
)

// WorkerOptions configures the application-hosted WorkerService.
//
// Zero values select local plaintext defaults. BindAddress controls the listening
// socket, while WorkerTarget is the address Dex advertises to running Flows; these
// commonly differ in containers or headless service deployments.
type WorkerOptions struct {
	// BindAddress is the plaintext WorkerService listener for Dex; it may differ from WorkerTarget. Default: ":8803".
	// It uses host:port syntax.
	BindAddress string
	// WorkerTarget is advertised to Dex and may differ from BindAddress. Default Address derives from BindAddress; Headless defaults false.
	WorkerTarget WorkerTarget
	// FlowServiceAddress is the plaintext Dex endpoint used for startup synchronization and blob hydration. Default: "localhost:8801".
	// Startup synchronization registers Attribute indexes.
	FlowServiceAddress string
	// AttributeIndexSyncTimeout bounds index registration at startup. Default: two minutes.
	// Negative durations are invalid.
	AttributeIndexSyncTimeout time.Duration
	// Logger defaults to the shared BlobCache logger.
	// It receives concurrent structured Worker logs and falls back to slog.Default.
	Logger Logger
}

// Worker hosts registered Flows over the private WorkerService protocol.
//
// A Worker is one-shot: construct it, call Start once, then call Stop during
// shutdown. Start blocks while serving. Step and RPC handlers may execute
// concurrently, so application handler state must provide its own synchronization.
//
//	worker, err := dex.NewWorker(registry, cache, dex.WorkerOptions{})
//	if err != nil {
//		return err
//	}
//	go func() { serveErr <- worker.Start() }()
//	defer worker.Stop(context.Background())
type Worker struct {
	registry                  *Registry
	bindAddress               string
	workerTarget              WorkerTarget
	grpcServer                *grpc.Server
	flowConn                  *grpc.ClientConn
	flowService               dexpb.FlowServiceClient
	attributeIndexSyncTimeout time.Duration
	logger                    Logger

	lifecycleMu sync.Mutex
	state       workerState
	done        chan struct{}
	finishOnce  sync.Once
}

type workerState uint8

const (
	workerCreated workerState = iota
	workerRunning
	workerStopping
	workerStopped
)

// NewWorker constructs a one-shot Worker from shared dependencies.
//
// registry supplies definitions and cache hydrates large values. It panics when
// registry or cache is nil. It validates listener, advertised target,
// FlowService address, and synchronization timeout, then creates the private gRPC
// server and FlowService connection without starting the listener. The returned
// Worker owns that connection; Start or Stop closes it.
func NewWorker(
	registry *Registry,
	cache *blobcache.Cache,
	options WorkerOptions,
) (*Worker, error) {
	if registry == nil {
		panic("dex.NewWorker requires Registry")
	}
	if cache == nil {
		panic("dex.NewWorker requires BlobCache")
	}
	bindAddress, target, flowServiceAddress, syncTimeout, err := resolveWorkerOptions(options)
	if err != nil {
		return nil, err
	}
	flowConn, err := grpc.NewClient(
		flowServiceAddress,
		grpc.WithTransportCredentials(insecure.NewCredentials()),
	)
	if err != nil {
		return nil, fmt.Errorf("dex: create FlowService client: %w", err)
	}

	logger := resolveLogger(options.Logger, cache.Logger())
	grpcServer := grpc.NewServer()
	flowService := dexpb.NewFlowServiceClient(flowConn)
	service := newWorkerService(
		registry,
		newValueHydrator(
			flowService,
			cache,
			logger,
		),
		logger,
	)
	dexpb.RegisterWorkerServiceServer(grpcServer, service)
	return &Worker{
		registry:                  registry,
		bindAddress:               bindAddress,
		workerTarget:              target,
		grpcServer:                grpcServer,
		flowConn:                  flowConn,
		flowService:               flowService,
		attributeIndexSyncTimeout: syncTimeout,
		logger:                    logger,
		state:                     workerCreated,
		done:                      make(chan struct{}),
	}, nil
}

func resolveWorkerOptions(
	options WorkerOptions,
) (string, WorkerTarget, string, time.Duration, error) {
	bindAddress := strings.TrimSpace(options.BindAddress)
	if bindAddress == "" {
		bindAddress = defaultWorkerBindAddress
	}
	bindHost, bindPort, err := validateWorkerBindAddress(bindAddress)
	if err != nil {
		return "", WorkerTarget{}, "", 0, err
	}

	target, err := resolveAdvertisedWorkerTarget(
		bindHost,
		bindPort,
		options.WorkerTarget,
	)
	if err != nil {
		return "", WorkerTarget{}, "", 0, err
	}
	flowServiceAddress := strings.TrimSpace(options.FlowServiceAddress)
	if flowServiceAddress == "" {
		flowServiceAddress = defaultFlowServiceTarget
	}
	if err := validatePlaintextTarget(flowServiceAddress, false); err != nil {
		return "", WorkerTarget{}, "", 0, fmt.Errorf(
			"dex: invalid FlowService address: %w",
			err,
		)
	}
	syncTimeout := options.AttributeIndexSyncTimeout
	if syncTimeout < 0 {
		return "", WorkerTarget{}, "", 0, fmt.Errorf(
			"dex: Attribute index sync timeout must be positive",
		)
	}
	if syncTimeout == 0 {
		syncTimeout = defaultAttributeIndexSyncTimeout
	}
	return bindAddress, target, flowServiceAddress, syncTimeout, nil
}

func validateWorkerBindAddress(address string) (string, string, error) {
	host, port, err := net.SplitHostPort(address)
	if err != nil {
		return "", "", fmt.Errorf("dex: invalid Worker bind address %q: %w", address, err)
	}
	portNumber, err := strconv.Atoi(port)
	if err != nil || portNumber < 1 || portNumber > 65535 {
		return "", "", fmt.Errorf("dex: invalid Worker bind address %q: port must be 1-65535", address)
	}
	return host, port, nil
}

func resolveAdvertisedWorkerTarget(
	bindHost string,
	bindPort string,
	target WorkerTarget,
) (WorkerTarget, error) {
	target.Address = strings.TrimSpace(target.Address)
	if target.Address == "" {
		targetHost := bindHost
		if targetHost == "" || targetHost == "0.0.0.0" || targetHost == "::" {
			targetHost = "localhost"
		}
		target.Address = net.JoinHostPort(targetHost, bindPort)
	}
	if err := validatePlaintextTarget(target.Address, target.Headless); err != nil {
		return WorkerTarget{}, fmt.Errorf("dex: invalid Worker target: %w", err)
	}
	return target, nil
}

func validatePlaintextTarget(address string, requireHostPort bool) error {
	trimmed := strings.TrimSpace(address)
	if trimmed != address {
		return fmt.Errorf("target address must not contain surrounding whitespace")
	}
	lower := strings.ToLower(trimmed)
	if lower == "" {
		return fmt.Errorf("target address must not be empty")
	}
	if strings.HasPrefix(lower, "http://") || strings.HasPrefix(lower, "https://") {
		return fmt.Errorf("target %q must use plaintext gRPC", address)
	}
	if requireHostPort {
		_, port, err := net.SplitHostPort(address)
		if err != nil {
			return fmt.Errorf("headless target %q must use host:port: %w", address, err)
		}
		portNumber, err := strconv.Atoi(port)
		if err != nil || portNumber < 1 || portNumber > 65535 {
			return fmt.Errorf("headless target %q must use port 1-65535", address)
		}
	}
	return nil
}

// Start serves WorkerService and blocks until Stop or a serve failure.
//
// Start may be called exactly once. Before listening, it synchronizes Attribute
// indexes and waits up to
// WorkerOptions.AttributeIndexSyncTimeout for Dex to accept the Registry's indexes.
// It returns nil after a normal Stop, or an error for synchronization, listener,
// serving, or invalid lifecycle state failures. Call Start on a dedicated goroutine
// when the application must continue doing other work.
func (worker *Worker) Start() error {
	worker.lifecycleMu.Lock()
	if worker.state != workerCreated {
		state := worker.state
		worker.lifecycleMu.Unlock()
		return fmt.Errorf("dex: Worker cannot start from state %s", state)
	}
	syncCtx, cancelSync := context.WithTimeout(context.Background(), worker.attributeIndexSyncTimeout)
	_, err := worker.flowService.SyncAttributeIndexes(syncCtx, &dexpb.SyncAttributeIndexRequest{
		AttributeIndexes: worker.registry.attributeIndexes,
	})
	cancelSync()
	if err != nil {
		worker.state = workerStopped
		worker.lifecycleMu.Unlock()
		worker.finish()
		return fmt.Errorf("dex: sync attribute indexes: %w", err)
	}
	listener, err := net.Listen("tcp", worker.bindAddress)
	if err != nil {
		worker.state = workerStopped
		worker.lifecycleMu.Unlock()
		worker.finish()
		return fmt.Errorf("dex: listen on %q: %w", worker.bindAddress, err)
	}
	worker.state = workerRunning
	worker.lifecycleMu.Unlock()

	serveErr := worker.grpcServer.Serve(listener)
	if errors.Is(serveErr, grpc.ErrServerStopped) {
		serveErr = nil
	}
	worker.lifecycleMu.Lock()
	worker.state = workerStopped
	worker.lifecycleMu.Unlock()
	worker.finish()
	if serveErr != nil {
		return fmt.Errorf("dex: serve WorkerService: %w", serveErr)
	}
	return nil
}

// Stop drains handlers until ctx expires, then force-stops the Worker.
//
// Stop accepts calls before Start and repeated calls after shutdown. A nil context is
// invalid. If ctx expires while handlers are draining, Stop force-closes the gRPC
// server, waits for cleanup, and returns ctx.Err. Stop closes the Worker's internal
// FlowService connection before returning.
func (worker *Worker) Stop(ctx context.Context) error {
	if ctx == nil {
		return fmt.Errorf("dex: Worker Stop context must not be nil")
	}

	worker.lifecycleMu.Lock()
	switch worker.state {
	case workerCreated:
		worker.state = workerStopped
		worker.lifecycleMu.Unlock()
		worker.grpcServer.Stop()
		worker.finish()
		return nil
	case workerRunning:
		worker.state = workerStopping
		go worker.grpcServer.GracefulStop()
	case workerStopping:
	case workerStopped:
		worker.lifecycleMu.Unlock()
		return nil
	}
	done := worker.done
	worker.lifecycleMu.Unlock()

	select {
	case <-done:
		return nil
	case <-ctx.Done():
		worker.grpcServer.Stop()
		<-done
		return ctx.Err()
	}
}

func (worker *Worker) finish() {
	worker.finishOnce.Do(func() {
		if err := worker.flowConn.Close(); err != nil {
			worker.logger.Error("close Worker FlowService connection", "error", err)
		}
		close(worker.done)
	})
}

// WorkerTarget returns a copy suitable for FlowConfig.WorkerTarget.
//
// The result is also suitable for ClientOptions.WorkerTarget. Mutating it does not
// reconfigure the Worker.
func (worker *Worker) WorkerTarget() *WorkerTarget {
	target := worker.workerTarget
	return &target
}

func (state workerState) String() string {
	switch state {
	case workerCreated:
		return "created"
	case workerRunning:
		return "running"
	case workerStopping:
		return "stopping"
	case workerStopped:
		return "stopped"
	default:
		return "unknown"
	}
}

File: sdk-python/dex/worker.py
# Legacy Materials in this file remain under their original licenses.
# See LEGACY_NOTICES.md.

# Modifications Copyright (c) 2026 Super Durable, Inc.
#
# Modifications after the Legacy Cutoff are licensed under the
# Super Durable Source License 1.0.
# Legacy Materials remain under their original licenses.
# See LICENSE and LEGACY_NOTICES.md.

from __future__ import annotations

import os
import threading
from concurrent.futures import ThreadPoolExecutor
from types import TracebackType

import grpc

from dex._value_hydrator import ValueHydrator
from dex._value_mapper import ValueMapper
from dex._worker_dispatcher import WorkerDispatcher
from dex._worker_service import WorkerService
from dex.blob_cache import BlobCache
from dex.dexpb import dex_pb2 as pb
from dex.dexpb import dex_pb2_grpc
from dex.flow import Registry
from dex.worker_options import WorkerOptions, WorkerTarget


class Worker:
    """Host synchronous registered Step and RPC handlers over WorkerService.

    A Worker is one-shot. Call ``start`` on a dedicated thread because it blocks,
    then call ``stop`` or ``close`` during shutdown. Handlers run concurrently in a
    bounded thread pool and must synchronize shared application state.

    Attributes:
        registry: The immutable Flow Registry served by this Worker.
        blob_cache: The shared cache used to hydrate large values.
        options: The effective WorkerOptions.

    Examples:
        >>> worker = Worker(registry, cache)
        >>> thread = threading.Thread(target=worker.start, daemon=True)
        >>> thread.start()
        >>> worker.worker_target.address
        'localhost:8803'
        >>> worker.stop()
    """

    def __init__(
        self,
        registry: Registry,
        blob_cache: BlobCache,
        options: WorkerOptions | None = None,
    ) -> None:
        """Construct a Worker without starting its listener.

        Args:
            registry: A Registry created for synchronous handlers.
            blob_cache: An open cache shared for value hydration.
            options: Networking and startup options; ``None`` uses defaults.

        Raises:
            ValueError: If a configured target or bind address is malformed.
        """
        self.registry = registry
        self.blob_cache = blob_cache
        self.options = options or WorkerOptions()
        self._lock = threading.Lock()
        self._state = "created"
        self._flow_channel = grpc.insecure_channel(self.options.server_address)
        self._flow_service = dex_pb2_grpc.FlowServiceStub(  # type: ignore[no-untyped-call]
            self._flow_channel
        )
        values = ValueMapper(registry.codec_registry)
        dispatcher = WorkerDispatcher(
            registry,
            values,
            ValueHydrator(self._flow_service, blob_cache),
        )
        concurrency = max(2, min(32, os.cpu_count() or 2))
        self._executor = ThreadPoolExecutor(
            max_workers=concurrency,
            thread_name_prefix="dex-python-handler",
        )
        self._server = grpc.server(
            self._executor,
            maximum_concurrent_rpcs=concurrency,
        )
        dex_pb2_grpc.add_WorkerServiceServicer_to_server(  # type: ignore[no-untyped-call]
            WorkerService(dispatcher),
            self._server,
        )
        self._bound_port = 0
        self._worker_target = self.options.worker_target or WorkerTarget(
            self._target_address(self.options.bind_address, 0)
        )

    def __enter__(self) -> Worker:
        """Return this Worker for synchronous context-manager use.

        Entering does not call ``start``.

        Returns:
            This Worker instance.
        """
        return self

    @property
    def worker_target(self) -> WorkerTarget:
        """Return the effective endpoint advertised to Dex.

        Before ``start``, an ephemeral bind port may not yet be resolved. After
        binding, the property contains the actual selected port.

        Returns:
            The immutable advertised WorkerTarget.
        """
        return self._worker_target

    def __exit__(
        self,
        exception_type: type[BaseException] | None,
        exception: BaseException | None,
        traceback: TracebackType | None,
    ) -> None:
        """Close the Worker when leaving a context-manager block.

        Args:
            exception_type: The active exception type, if any.
            exception: The active exception value, if any.
            traceback: The active traceback, if any.
        """
        self.close()

    def start(self) -> None:
        """Synchronize Attribute indexes, serve WorkerService, and block.

        ``start`` may be called exactly once. It first contacts FlowService using the
        configured synchronization timeout, then binds the listener and waits until
        ``stop`` shuts the server down.

        Raises:
            RuntimeError: If lifecycle state, index synchronization, or binding fails.
        """
        with self._lock:
            if self._state != "created":
                raise RuntimeError(f"Worker cannot start from state {self._state}")
            try:
                self._flow_service.SyncAttributeIndexes(
                    pb.SyncAttributeIndexRequest(
                        attribute_indexes=dict(self.registry._attribute_indexes)
                    ),
                    timeout=self.options.attribute_index_sync_timeout.total_seconds(),
                )
            except grpc.RpcError as failure:
                self._state = "stopped"
                self._flow_channel.close()
                self._executor.shutdown(wait=True, cancel_futures=True)
                raise RuntimeError("cannot synchronize Attribute indexes") from failure
            self._bound_port = self._server.add_insecure_port(self.options.bind_address)
            if self._bound_port == 0:
                self._state = "stopped"
                self._flow_channel.close()
                self._executor.shutdown(wait=True, cancel_futures=True)
                raise RuntimeError(
                    f"cannot bind Python Worker to {self.options.bind_address}"
                )
            if self.options.worker_target is None:
                self._worker_target = WorkerTarget(
                    self._target_address(self.options.bind_address, self._bound_port)
                )
            self._state = "running"
            self._server.start()
        self._server.wait_for_termination()

    def stop(self) -> None:
        """Drain handlers for five seconds and release Worker resources.

        Calls before ``start`` and repeated calls after shutdown are safe. Remaining
        handler futures are canceled after the gRPC grace period.
        """
        with self._lock:
            if self._state in ("stopped", "closed"):
                return
            self._state = "stopping"
        self._server.stop(grace=5).wait(timeout=10)
        self._flow_channel.close()
        self._executor.shutdown(wait=True, cancel_futures=True)
        with self._lock:
            if self._state != "closed":
                self._state = "stopped"

    def close(self) -> None:
        """Stop the Worker and permanently mark it closed.

        ``close`` is idempotent; a closed Worker cannot be started again.
        """
        self.stop()
        with self._lock:
            self._state = "closed"

    @staticmethod
    def _target_address(bind_address: str, bound_port: int) -> str:
        host, separator, port = bind_address.rpartition(":")
        if not separator or not port:
            raise ValueError("Worker bind address requires a port")
        if bound_port == 0:
            bound_port = int(port)
        if host in ("", "0.0.0.0", "::", "[::]"):
            host = "localhost"
        return f"{host}:{bound_port}"
